### ISSUE TITLE
Fetch hook doesn't always get applied before initial load

### ISSUE THREAD

**Issue by lewisl9029** [guybedford/es-module-shims#108]
I recently ran into a bug where `importShim.load()` would very occasionally get called before a subsequent script has the chance to set the fetch script in a new project I'm working on.

I've observed that it generally happens when the subsequent script ends up taking a significant amount of time to load compared to es-module-shims: 
![image](https://user-images.githubusercontent.com/6934200/103046326-b14e8700-453c-11eb-8c10-81a2e97d763b.png)

Here's what my script loading order looks like: 

```
  <body>
    <link href="/index.css" rel="stylesheet" />
    <noscript>You need to enable JavaScript to run this app.</noscript>
    <div id="root"></div>
    <script type="module" src="/es-module-shims.js"></script>
    <script type="module" src="/dev.js"></script>
    <script type="importmap-shim" src="/production.importmap"></script>
    <script type="module-shim" src="/index.js"></script>
  </body>
```

I've forked the jspm.dev bundle of es-module-shims to a local file and added a log statement (`processing scripts
`) at https://github.com/guybedford/es-module-shims/blob/master/src/es-module-shims.js#L218

And added a log at the beginning of the dev.js file that sets the fetch hook (`setting fetch shim`).

This is what the logs look like when the error occurs:
![image](https://user-images.githubusercontent.com/6934200/103046477-4a7d9d80-453d-11eb-869b-f0889e657650.png)

This is what the logs normally look like:
![image](https://user-images.githubusercontent.com/6934200/103046559-8ca6df00-453d-11eb-80c7-4e39e1be70f0.png)

Note how the order between `setting fetch shim` and `processing scripts` is reversed.

I also tried to change https://github.com/guybedford/es-module-shims/blob/master/src/es-module-shims.js#L213 to `setTimeout(processScripts, 0)` as per this discussion: https://github.com/guybedford/es-module-shims/pull/95/files#r500766794

But it didn't seem to have much of an effect on the frequency of occurrence, possibly because the loading of dev.js took longer than the next async tick.

Would it make sense to provide some way for users to provide the fetch hook before loading es-module-shims (i.e. so I can swap the loading order of dev.js and es-module-shims here)? Happy to try out any other solutions folks might have in mind as well.

**Comment by guybedford** [comment]
In the example above, is your fetch hook defined in the `/dev.js` file?

If so, the issue might be that we are calling the first `processScripts` synchronously, and simply having it called after a tick via a `Promise.resolve().then(processScripts)` or a `setTimeout(processScripts, 1)` would avoid the issue?

If you'd like to work on a PR that would be very welcome.

**Comment by lewisl9029** [comment]
@guybedford thanks for following up. Yep the fetch hook is defined in `/dev.js`

I did actually try `setTimeout(processScripts, 0)` as suggested here: https://github.com/guybedford/es-module-shims/pull/95/files#r500766794

But the issue remained. I haven't tried Promise.resolve or setting timeout to 1 yet but those feel somewhat unlikely to behave any differently. 

The only real solutions I can think of at the moment is either making es-module-shims itself read from some global to apply a user-defined fetch hook that's loaded before es-module-shims, or providing an alternative entry point for es-module-shims where it doesn't self-initialize on load, and instead exposes some init function for users to call at the appropriate timing.

If we can settle on an approach I'd be happy to submit a PR to implement. Open to other suggestions of course.

**Comment by guybedford** [comment]
The other option might be to only do script processing on domready, but that feels far too late in the loading cycle for real use cases.

Yes, perhaps the easiest is to define a `__esmsFetch` global to check which takes priority over `fetch`.

**Comment by guybedford** [comment]
Released in 0.9.0.
